pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java

```diff
@@ -1570,6 +1577,7 @@ protected CompletableFuture<Optional<Topic>> loadOrCreatePersistentTopic(final S
                         }
                     }
                 }).exceptionally(ex -> {
+                    pulsar.getExecutor().execute(() -> topics.remove(topic));
```

Comment:
We have a lot of places to remove the Future from the topics map.
Actually, we can only handle it at the place that put the Future to the topics map.

The principles are 

- For the places complete the future, just complete the future with a value or an exception
- Use the exceptionally() method to handle the failed future and always use topics.remove(topic, future) to avoid any race conditions.

For example:

```java
return getTopicPoliciesBypassSystemTopic(topicName).exceptionally(ex -> {
      final Throwable rc = FutureUtil.unwrapCompletionException(ex);
      final String errorInfo = String.format("Topic creation encountered an exception by initialize"
              + " topic policies service. topic_name=%s error_message=%s", topicName,
              rc.getMessage());
      log.error(errorInfo, rc);
      throw FutureUtil.wrapToCompletionException(new ServiceUnitNotReadyException(errorInfo));
  }).thenCompose(optionalTopicPolicies -> {
      final TopicPolicies topicPolicies = optionalTopicPolicies.orElse(null);
      CompletableFuture<Optional<Topic>> future = topics.computeIfAbsent(topicName.toString(), (tpName) -> {
          if (topicName.isPartitioned()) {
              final TopicName topicNameEntity = TopicName.get(topicName.getPartitionedTopicName());
              return fetchPartitionedTopicMetadataAsync(topicNameEntity)
                      .thenCompose((metadata) -> {
                          // Allow crate non-partitioned persistent topic that name includes
                          // `partition`
                          if (metadata.partitions == 0
                                  || topicName.getPartitionIndex() < metadata.partitions) {
                              return loadOrCreatePersistentTopic(tpName, createIfMissing,
                                      properties, topicPolicies);
                          }
                          final String errorMsg =
                                  String.format("Illegal topic partition name %s with max allowed "
                                          + "%d partitions", topicName, metadata.partitions);
                          log.warn(errorMsg);
                          return FutureUtil.failedFuture(
                                  new BrokerServiceException.NotAllowedException(errorMsg));
                      });
          }
          return loadOrCreatePersistentTopic(tpName, createIfMissing, properties, topicPolicies);
      }).thenCompose(optionalTopic -> {
          if (!optionalTopic.isPresent() && createIfMissing) {
              log.warn("[{}] Try to recreate the topic with createIfMissing=true "
                      + "but the returned topic is empty", topicName);
              return getTopic(topicName, createIfMissing, properties);
          }
          return CompletableFuture.completedFuture(optionalTopic);
      });
      future.whenComplete((__, ex) -> {
          if (ex != null) {
              pulsar.getExecutor().execute(() -> topics.remove(topicName.toString(), future));
          }
      });
      return future;
  });
```

Reply:
Good point, I changed it, PTAL.